sympy/simplify/simplify.py

```diff
@@ -315,18 +317,79 @@ def hypersimp(f, k):
     """
     f = sympify(f)
 
-    g = f.subs(k, k + 1) / f
+    if not f.is_commutative:
+        return None
 
-    g = g.rewrite(gamma)
-    if g.has(Piecewise):
-        g = piecewise_fold(g)
-        g = g.args[-1][0]
-    g = expand_func(g)
-    g = powsimp(g, deep=True, combine='exp')
+    f = f.rewrite(gamma)
+
+    f = expand_func(f)
+    f = factor_terms(f)
+
+    return _get_term_ratio(f, k)
+
+
+def _get_term_ratio(f, k):
+    if f.is_rational_function(k):
+        # If f is a rational function then its term ratio is as well
+        return together(f.subs(k, k+1) / f)
+
+    elif isinstance(f, Mul):
+        # Compute term ratio for each factor
+        ratios = []
+        for a in f.args:
+            r = _get_term_ratio(a, k)
+            if r is None:
+                return None
+            ratios.append(r)
+        return Mul(*ratios)
+
+    elif isinstance(f, (Pow, exp)):
+        b, e = f.as_base_exp()
+        if e.is_Integer:
+            # f(k)**e, Compute term ratio for f(k)
+            r = _get_term_ratio(b, k)
+            if r is not None:
+                return r**e
+        elif not b.has_xfree({k}):
+            # b**(c*k + ...), ratio is b**c
+            c = _linear_coeff(e, k)
+            if c is not None:
+                return b**c
+
+    elif isinstance(f, gamma):
+        a = f.args[0]
+        c = _linear_coeff(a, k)
+        if c is not None and c.is_Integer:
+            # gamma(a) = gamma(c*k + ...)
+            return RisingFactorial(a, c)
 
-    if g.is_rational_function(k):
-        return simplify(g, ratio=S.Infinity)
     else:
+        g = f.subs(k, k+1)/f
+        if g.is_rational_function(k):
+            return together(g)
+
+        elif isinstance(f, Piecewise):
```

Comment:
Testing `isinstance(f, Piecewise)` misses cases where the expression contains a Piecewise:
```python
In [11]: p = 1 + Piecewise((2**k-1, x > 0), (2**(k-1)-1, True))

In [12]: p
Out[12]:
⎛⎧   k                 ⎞
⎜⎪  2  - 1    for x > 0⎟
⎜⎨                     ⎟ + 1
⎜⎪ k - 1               ⎟
⎝⎩2      - 1  otherwise⎠

In [13]: hypersimp(p, k)

In [14]: hypersimp(piecewise_fold(p), k)
Out[14]: 2
```